### PR TITLE
Raise exceptions in mapper

### PR DESCRIPTION
**Pull request by carolineechen**
return error messages in mapper.map output instead of returning None. update retry logic to handle this.

### CODE REVIEW

**Comment by carolineechen** [comment]
* **#605** <a href="https://app.graphite.dev/github/pr/run-house/runhouse/605?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @carolineechen and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: runhouse/resources/functionals/mapper.py [hunk 1]
```diff
@@ -159,7 +165,10 @@ def call_method_on_replica(job):
                 results.append(
                     list(
                         tqdm(
-                            executor.map(call_method_on_replica, jobs), total=len(jobs)
+                            executor.submit(
+                                call_method_on_replica, jobs, i != retries - 1
```
Comment (rohinb2) [inline runhouse/resources/functionals/mapper.py:169]: what is going on here lol

Comment (carolineechen) [reply runhouse/resources/functionals/mapper.py]: added a retry field to `call_module_on_replica` to handles exceptions differently depending on if it's retrying again or not.

File: runhouse/resources/functionals/mapper.py [hunk 2]
```diff
@@ -148,18 +151,23 @@ def call_method_on_replica(job):
...
+                for job in jobs
+            ]
             for fut in tqdm(concurrent.futures.as_completed(futs), total=len(jobs)):
                 results.extend([fut.result()])
             for i in range(retries):
                 if len(retry_list) == 0:
                     break
                 logger.info(f"Retry {i}: {len(retry_list)} failed jobs")
                 jobs, retry_list = retry_list, []
+                retry = i != retries - 1
```
Comment (dongreenberg) [inline runhouse/resources/functionals/mapper.py:165]: Oh this is making it raise if we're on the last retry? Nice

**Review by dongreenberg — APPROVED** [review]
Looks great!

**Pull request merged**